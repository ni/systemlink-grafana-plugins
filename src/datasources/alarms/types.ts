import { DataQuery } from '@grafana/schema';

export interface AlarmsQuery extends DataQuery {}
