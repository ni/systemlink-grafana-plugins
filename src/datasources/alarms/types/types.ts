import { DataQuery } from '@grafana/schema';

export interface AlarmsQuery extends DataQuery {
  queryType?: QueryType;
}

export interface AlarmsVariableQuery extends AlarmsQuery {
  filter?: string;
  descending?: boolean;
  take?: number;
}

export enum QueryType {
  AlarmsCount = 'Alarms Count',
  ListAlarms = 'List Alarms'
}

export interface QueryAlarmsRequest {
  filter?: string;
  transitionInclusionOption?: TransitionInclusionOption;
  take?: number;
  orderBy?: string;
  orderByDescending?: boolean;
  continuationToken?: string;
  returnCount?: boolean;
  returnMostRecentlyOccurredOnly?: boolean;
}

export interface QueryAlarmsResponse {
  alarms: Alarm[];
  totalCount?: number;
  continuationToken?: string;
}

export interface Alarm {
  instanceId: string;
  alarmId: string;
  workspace: string;
  active: boolean;
  clear: boolean;
  acknowledged: boolean;
  acknowledgedAt: string | null;
  acknowledgedBy: string | null;
  occurredAt: string;
  updatedAt: string;
  createdBy: string;
  transitions: AlarmTransition[];
  transitionOverflowCount: number;
  currentSeverityLevel: number;
  highestSeverityLevel: number;
  mostRecentSetOccurredAt: string | null;
  mostRecentTransitionOccurredAt: string | null;
  channel: string;
  condition: string;
  displayName: string;
  description: string;
  keywords: string[];
  properties: {
    [key: string]: string;
  };
  resourceType: string;
}

enum TransitionInclusionOption {
  None = 'NONE',
  MostRecentOnly = 'MOST_RECENT_ONLY',
  All = 'ALL',
};

interface AlarmTransition {
  transitionType: AlarmTransitionType;
  occurredAt: string;
  severityLevel: number;
  value: string;
  condition: string;
  shortText: string;
  detailText: string;
  keywords: string[];
  properties: {
    [key: string]: string;
  };
}

export enum AlarmTransitionType {
  Clear = 'CLEAR',
  Set = 'SET',
};
