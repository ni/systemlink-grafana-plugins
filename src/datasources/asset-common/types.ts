
export interface AssetsResponse {
    assets: AssetModel[],
    totalCount: number
  }
  
  export interface AssetLocationModel {
    minionId: string,
    parent: string,
    resourceUri: string,
    slotNumber: number,
    state: AssetPresenceWithSystemConnectionModel,
    physicalLocation?: string
  }
  
  export interface AssetPresenceWithSystemConnectionModel {
    assetPresence: "INITIALIZING" | "UNKNOWN" | "NOT_PRESENT" | "PRESENT" | string,
    // to be compatible with both SLS and SLE
    systemConnection?: "APPROVED" | "DISCONNECTED" | "CONNECTED_UPDATE_PENDING" | "CONNECTED_UPDATE_SUCCESSFUL" | "CONNECTED_UPDATE_FAILED" | "UNSUPPORTED" | "ACTIVATED" | "CONNECTED" | string
  }
  
  export interface AssetModel {
    modelName: string,
    modelNumber: number,
    serialNumber: string,
    vendorName: string,
    vendorNumber: number,
    busType: 'BUILT_IN_SYSTEM' | 'PCI_PXI' | 'USB' | 'GPIB' | 'VXI' | 'SERIAL' | 'TCP_IP' | 'CRIO' | 'SCXI' | 'CDAQ' | 'SWITCH_BLOCK' | 'SCC' | 'FIRE_WIRE' | 'ACCESSORY' | 'CAN' | 'SWITCH_BLOCK_DEVICE' | 'SLSC' | string,
    name: string,
    assetType: 'GENERIC' | 'DEVICE_UNDER_TEST' | 'FIXTURE' | 'SYSTEM' | string,
    firmwareVersion: string,
    hardwareVersion: string,
    visaResourceName: string,
    temperatureSensors: any[],
    supportsSelfCalibration: boolean,
    supportsExternalCalibration: boolean,
    customCalibrationInterval?: number,
    selfCalibration?: any,
    isNIAsset: boolean,
    workspace: string,
    fileIds: string[],
    supportsSelfTest: boolean,
    supportsReset: boolean,
    id: string,
    location: AssetLocationModel,
    calibrationStatus: 'OK' | 'APPROACHING_RECOMMENDED_DUE_DATE' | 'PAST_RECOMMENDED_DUE_DATE' | string,
    isSystemController: boolean,
    externalCalibration?: ExternalCalibrationModel,
    discoveryType: 'MANUAL' | 'AUTOMATIC' | string,
    properties: Record<string, string>,
    keywords: string[],
    lastUpdatedTimestamp: string,
  }
  
  export interface ExternalCalibrationModel {
    temperatureSensors: any[],
    isLimited?: boolean,
    date: string,
    recommendedInterval: number,
    nextRecommendedDate: string,
    nextCustomDueDate?: string,
    comments: string,
    entryType: "AUTOMATIC" | "MANUAL" | string,
    operator: ExternalCalibrationOperatorModel
  }
  
  export interface ExternalCalibrationOperatorModel {
    displayName: string;
    userId: string
  }
