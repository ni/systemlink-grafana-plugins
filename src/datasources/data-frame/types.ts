import { DataQuery } from '@grafana/schema';

export interface DataFrameQuery extends DataQuery {
  tableId?: string;
  columns?: string[];
  decimationMethod?: string;
  filterNulls?: boolean;
  applyTimeFilters?: boolean;
}

export const defaultQuery: Omit<ValidDataFrameQuery, 'refId'> = {
  tableId: '',
  columns: [],
  decimationMethod: 'LOSSY',
  filterNulls: false,
  applyTimeFilters: false
};

export type ValidDataFrameQuery = DataFrameQuery & Required<Omit<DataFrameQuery, keyof DataQuery>>;

export type ColumnDataType = 'BOOL' | 'INT32' | 'INT64' | 'FLOAT32' | 'FLOAT64' | 'STRING' | 'TIMESTAMP';

export interface Column {
  name: string;
  dataType: ColumnDataType;
  columnType: 'INDEX' | 'NULLABLE' | 'NORMAL';
  properties: Record<string, string>;
}

export interface ColumnFilter {
  column: string;
  operation:
    | 'EQUALS'
    | 'LESS_THAN'
    | 'LESS_THAN_EQUALS'
    | 'GREATER_THAN'
    | 'GREATER_THAN_EQUALS'
    | 'NOT_EQUALS'
    | 'CONTAINS'
    | 'NOT_CONTAINS';
  value: string | null;
}

export interface TableMetadata {
  columns: Column[];
  id: string;
  name: string;
  workspace: string;
}

export interface TableMetadataList {
  tables: TableMetadata[];
  continuationToken: string;
}

export interface TableDataRows {
  frame: { columns: string[]; data: string[][] };
  continuationToken: string;
}

export interface SystemLinkError {
  error: {
    args: string[];
    code: number;
    message: string;
    name: string;
  }
}

export function isSystemLinkError(error: any): error is SystemLinkError {
  return Boolean(error?.error?.code) && Boolean(error?.error?.name);
}
