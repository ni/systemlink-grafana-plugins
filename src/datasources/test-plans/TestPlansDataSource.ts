import { DataFrameDTO, DataQueryRequest, DataSourceInstanceSettings, FieldType, LegacyMetricFindQueryOptions, MetricFindValue, TestDataSourceResponse } from '@grafana/data';
import { BackendSrv, TemplateSrv, getBackendSrv, getTemplateSrv } from '@grafana/runtime';
import { DataSourceBase } from 'core/DataSourceBase';
import { Asset, OrderByOptions, OutputType, Projections, Properties, PropertiesProjectionMap, QueryTemplatesResponse, QueryTestPlansResponse, TemplateResponseProperties, TestPlanResponseProperties, TestPlansQuery, TestPlansVariableQuery } from './types';
import { getWorkspaceName, getVariableOptions, queryInBatches } from 'core/utils';
import { QueryBuilderOption, QueryResponse } from 'core/types';
import { isTimeField, transformDuration } from './utils';
import { QUERY_TEMPLATES_BATCH_SIZE, QUERY_TEMPLATES_REQUEST_PER_SECOND, QUERY_TEST_PLANS_MAX_TAKE, QUERY_TEST_PLANS_REQUEST_PER_SECOND } from './constants/QueryTestPlans.constants';
import { AssetUtils } from './asset.utils';
import { WorkspaceUtils } from 'shared/workspace.utils';
import { SystemUtils } from 'shared/system.utils';
import { QueryBuilderOperations } from 'core/query-builder.constants';
import { ExpressionTransformFunction, transformComputedFieldsQuery } from 'core/query-builder.utils';
import { UsersUtils } from 'shared/users.utils';

export class TestPlansDataSource extends DataSourceBase<TestPlansQuery> {
  constructor(
    readonly instanceSettings: DataSourceInstanceSettings,
    readonly backendSrv: BackendSrv = getBackendSrv(),
    readonly templateSrv: TemplateSrv = getTemplateSrv()
  ) {
    super(instanceSettings, backendSrv, templateSrv);
    this.assetUtils = new AssetUtils(instanceSettings, backendSrv);
    this.workspaceUtils = new WorkspaceUtils(this.instanceSettings, this.backendSrv);
    this.systemUtils = new SystemUtils(instanceSettings, backendSrv);
    this.usersUtils = new UsersUtils(instanceSettings, backendSrv);
  }

  baseUrl = `${this.instanceSettings.url}/niworkorder/v1`;
  queryTestPlansUrl = `${this.baseUrl}/query-testplans`;
  queryTemplatesUrl = `${this.baseUrl}/query-testplan-templates`;
  assetUtils: AssetUtils;
  workspaceUtils: WorkspaceUtils;
  systemUtils: SystemUtils;
  usersUtils: UsersUtils;

  defaultQuery = {
    outputType: OutputType.Properties,
    properties: [
      Properties.NAME,
      Properties.STATE,
      Properties.ASSIGNED_TO,
      Properties.PRODUCT,
      Properties.DUT_ID,
      Properties.PLANNED_START_DATE_TIME,
      Properties.ESTIMATED_DURATION_IN_SECONDS,
      Properties.SYSTEM_NAME,
      Properties.UPDATED_AT
    ] as Properties[],
    orderBy: OrderByOptions.UPDATED_AT,
    descending: true,
    recordCount: 1000
  };

  readonly globalVariableOptions = (): QueryBuilderOption[] => getVariableOptions(this);

  async runQuery(query: TestPlansQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
    const workspaces = await this.workspaceUtils.getWorkspaces();
    const systemAliases = await this.systemUtils.getSystemAliases();
    const users = await this.usersUtils.getUsers();

    if (query.queryBy) {
      query.queryBy = transformComputedFieldsQuery(
        this.templateSrv.replace(query.queryBy, options.scopedVars),
        this.testPlansComputedDataFields
      );
    }

    if (query.outputType === OutputType.Properties) {
      const projectionAndFields = query.properties?.map(property => PropertiesProjectionMap[property]);
      const projection = [...new Set(projectionAndFields?.map(data => data.projection).flat()), Projections.ID];

      const testPlans = (
        await this.queryTestPlansInBatches(
          query.queryBy,
          query.orderBy,
          projection,
          query.recordCount,
          query.descending,
          true
        )).testPlans;

      if (testPlans.length > 0) {
        const labels = projectionAndFields?.map(data => data.label) ?? [];
        const fixtureNames = await this.getFixtureNames(labels, testPlans);
        const duts = await this.getDuts(labels, testPlans);
        const workOrderIdAndName = this.getWorkOrderIdAndName(labels, testPlans);
        const templatesName = await this.getTemplateNames(labels, testPlans);

        const fields = projectionAndFields?.map((data) => {
          const label = data.label;
          const field = data.field;
          const fieldType = isTimeField(field)
            ? FieldType.time
            : FieldType.string;
          const values = testPlans
            .map(data => data[field as unknown as keyof TestPlanResponseProperties] as any);

          // TODO: AB#3133188 Add support for other field mapping
          const fieldValues = values.map(value => {
            switch (label) {
              case PropertiesProjectionMap.FIXTURE_NAMES.label:
                const names = value.map((id: string) => fixtureNames.find(data => data.id === id)?.name);
                return names ? names.filter((name: string) => name !== '').join(', ') : value;
              case PropertiesProjectionMap.DUT_ID.label:
                const dutName = duts.find(data => data.id === value);
                return dutName ? dutName.name : value;
              case PropertiesProjectionMap.DUT_SERIAL_NUMBER.label:
                const dutSerial = duts.find(data => data.id === value);
                return dutSerial ? dutSerial.serialNumber : value;
              case PropertiesProjectionMap.WORKSPACE.label:
                const workspace = workspaces.get(value);
                return workspace ? getWorkspaceName([workspace], value) : value;
              case PropertiesProjectionMap.WORK_ORDER.label:
                const workOrder = workOrderIdAndName.find(data => data.id === value);
                const workOrderId = value ? `(${value})` : '';
                const workOrderName = (workOrder && workOrder?.name) ? workOrder.name : '';
                return `${workOrderName} ${workOrderId}`;
              case PropertiesProjectionMap.TEMPLATE.label:
                const template = templatesName.find(data => data.id === value);
                return template ? `${template.name} (${template.id})` : value;
              case PropertiesProjectionMap.ESTIMATED_DURATION_IN_SECONDS.label:
                return value ? transformDuration(value) : '';
              case PropertiesProjectionMap.SYSTEM_NAME.label:
                const system = systemAliases.get(value);
                return system ? system.alias : value;
              case PropertiesProjectionMap.ASSIGNED_TO.label:
              case PropertiesProjectionMap.CREATED_BY.label:
              case PropertiesProjectionMap.UPDATED_BY.label:
                const user = users.get(value);
                return user ? UsersUtils.getUserFullName(user) : '';
              case PropertiesProjectionMap.PROPERTIES.label:
                return value == null ? '' : JSON.stringify(value);
              default:
                return value == null ? '' : value;
            }
          });

          return {
            name: label,
            values: fieldValues,
            type: fieldType
          };
        });
        return {
          refId: query.refId,
          name: query.refId,
          fields: fields ?? [],
        };
      }
      return {
        refId: query.refId,
        name: query.refId,
        fields: [],
      };
    } else {
      const responseData = await this.queryTestPlans(
        query.queryBy,
        query.orderBy,
        undefined,
        0,
        undefined,
        undefined,
        true
      );

      return {
        refId: query.refId,
        name: query.refId,
        fields: [{ name: 'Total count', values: [responseData.totalCount] }],
      };
    }
  }

  shouldRunQuery(query: TestPlansQuery): boolean {
    return true;
  }

  private async getFixtureNames(labels: string[], testPlans: TestPlanResponseProperties[]): Promise<Asset[]> {
    if (labels.find(label => label === PropertiesProjectionMap.FIXTURE_NAMES.label)) {
      const fixtureIds = testPlans
        .map(data => data['fixtureIds'] as string[])
        .filter(data => data.length > 0)
        .flat();
      return await this.assetUtils.queryAssetsInBatches(fixtureIds);
    }
    return [];
  }

  private async getDuts(labels: string[], testPlans: TestPlanResponseProperties[]): Promise<Asset[]> {
    if (labels.find(label =>
      label === PropertiesProjectionMap.DUT_ID.label
      || label === PropertiesProjectionMap.DUT_SERIAL_NUMBER.label
    )) {
      const dutIds = testPlans
        .map(data => data['dutId'] as string)
        .filter(data => data != null);
      return await this.assetUtils.queryAssetsInBatches(dutIds);
    }
    return [];
  }

  private getWorkOrderIdAndName(labels: string[], testPlans: TestPlanResponseProperties[]): Array<{ id: string; name: string }> {
    if (labels.find(label => label === PropertiesProjectionMap.WORK_ORDER.label)) {
      return testPlans
        .map(data => ({
          id: data['workOrderId'] as string,
          name: data['workOrderName'] as string
        }))
        .filter(data => data.id != null);
    }
    return [];
  }

  private async getTemplateNames(labels: string[], testPlans: TestPlanResponseProperties[]): Promise<TemplateResponseProperties[]> {
    if (labels.find(label => label === PropertiesProjectionMap.TEMPLATE.label)) {
      const templateIds = testPlans
        .map(data => data['templateId'] as string)
        .filter(id => id != null);
      return await this.queryTestPlanTemplatesInBatches(templateIds);
    }
    return [];
  }

  async metricFindQuery(query: TestPlansVariableQuery, options: LegacyMetricFindQueryOptions): Promise<MetricFindValue[]> {
    const filter = query.queryBy ?
      transformComputedFieldsQuery(
        this.templateSrv.replace(query.queryBy, options.scopedVars),
        this.testPlansComputedDataFields
      )
      : undefined;

    const metadata = (await this.queryTestPlansInBatches(
      filter,
      query.orderBy,
      [Projections.ID, Projections.NAME],
      query.recordCount,
      query.descending
    )).testPlans;
    return metadata ? metadata.map(frame => ({ text: `${frame.name} (${frame.id})`, value: frame.id })) : [];
  }

  readonly testPlansComputedDataFields = new Map<string, ExpressionTransformFunction>(
    Object.values(PropertiesProjectionMap).map(({ field, projection }) => {
      const fieldName = field;
      const isTime = isTimeField(projection[0]);
      return [fieldName, isTime ? this.timeFieldsQuery(fieldName) : this.multipleValuesQuery(fieldName)];
    })
  );

  protected multipleValuesQuery(field: string): ExpressionTransformFunction {
    return (value: string, operation: string, _options?: any) => {
      const isMultiSelect = this.isMultiSelectValue(value);
      const valuesArray = this.getMultipleValuesArray(value);
      const logicalOperator = this.getLogicalOperator(operation);

      return isMultiSelect
        ? `(${valuesArray.map(val => `${field} ${operation} "${val}"`).join(` ${logicalOperator} `)})`
        : `${field} ${operation} "${value}"`;
    };
  }

  protected timeFieldsQuery(field: string): ExpressionTransformFunction {
    return (value: string, operation: string): string => {
      const formattedValue = value === '${__now:date}' ? new Date().toISOString() : value;
      return `${field} ${operation} "${formattedValue}"`;
    };
  }

  private isMultiSelectValue(value: string): boolean {
    return value.startsWith('{') && value.endsWith('}');
  }

  private getMultipleValuesArray(value: string): string[] {
    return value.replace(/({|})/g, '').split(',');
  }

  private getLogicalOperator(operation: string): string {
    return operation === QueryBuilderOperations.EQUALS.name ? '||' : '&&';
  }

  async testDatasource(): Promise<TestDataSourceResponse> {
    await this.post(this.queryTestPlansUrl, { take: 1 });
    return { status: 'success', message: 'Data source connected and authentication successful!' };
  }

  async queryTestPlansInBatches(
    filter?: string,
    orderBy?: string,
    projection?: Projections[],
    take?: number,
    descending = false,
    returnCount = false
  ): Promise<QueryTestPlansResponse> {
    const queryRecord = async (currentTake: number, token?: string): Promise<QueryResponse<TestPlanResponseProperties>> => {
      const response = await this.queryTestPlans(
        filter,
        orderBy,
        projection,
        currentTake,
        descending,
        token,
        returnCount
      );

      return {
        data: response.testPlans,
        continuationToken: response.continuationToken,
        totalCount: response.totalCount
      };
    };

    const batchQueryConfig = {
      maxTakePerRequest: QUERY_TEST_PLANS_MAX_TAKE,
      requestsPerSecond: QUERY_TEST_PLANS_REQUEST_PER_SECOND
    };

    const response = await queryInBatches(queryRecord, batchQueryConfig, take);

    return {
      testPlans: response.data,
      continuationToken: response.continuationToken,
      totalCount: response.totalCount
    };
  }

  async queryTestPlans(
    filter?: string,
    orderBy?: string,
    projection?: Projections[],
    take?: number,
    descending?: boolean,
    continuationToken?: string,
    returnCount = false
  ): Promise<QueryTestPlansResponse> {
    try {
      const response = await this.post<QueryTestPlansResponse>(this.queryTestPlansUrl, {
        filter,
        orderBy,
        descending,
        projection,
        take,
        continuationToken,
        returnCount
      });
      return response;
    } catch (error) {
      throw new Error(`An error occurred while querying test plans: ${error} `);
    }
  }

  async queryTestPlanTemplatesInBatches(
    templateIds: string[]
  ): Promise<TemplateResponseProperties[]> {
    const queryRecord = async (idsChunk: string[]): Promise<TemplateResponseProperties[]> => {
      return await this.queryTestPlanTemplates(idsChunk);
    };

    const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
    const response: TemplateResponseProperties[] = [];
    while (templateIds.length > 0) {
      const start = Date.now();

      for (let i = 0; i < QUERY_TEMPLATES_REQUEST_PER_SECOND && templateIds.length > 0; i++) {
        const idsToQuery = templateIds.splice(0, QUERY_TEMPLATES_BATCH_SIZE);

        try {
          const template = await queryRecord(idsToQuery);
          response.push(...template);
        } catch (error) {
          console.error(`Error fetching test plan templates:`, error);
        }
      }

      const elapsed = Date.now() - start;
      if (templateIds.length > 0 && elapsed < 1000) {
        await delay(1000 - elapsed);
      }
    }

    return response;
  }

  async queryTestPlanTemplates(templateIds: string[]): Promise<TemplateResponseProperties[]> {
    try {
      const filter = templateIds.map(id => `id = "${id}"`).join(' || ');
      const response = await this.post<QueryTemplatesResponse>(this.queryTemplatesUrl, {
        filter
      });
      return response.testPlanTemplates;
    } catch (error) {
      throw new Error(`An error occurred while querying test plan templates: ${error} `);
    }
  }
}
