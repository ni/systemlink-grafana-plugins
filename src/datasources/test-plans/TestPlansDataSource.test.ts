import { MockProxy } from "jest-mock-extended";
import { TestPlansDataSource } from "./TestPlansDataSource";
import { BackendSrv } from "@grafana/runtime";
import { createFetchError, createFetchResponse, requestMatching, setupDataSource } from "test/fixtures";
import { OrderByOptions, OutputType, Projections, Properties, QueryTestPlansResponse, TestPlansVariableQuery } from "./types";
import { DataQueryRequest, LegacyMetricFindQueryOptions } from "@grafana/data";

let datastore: TestPlansDataSource, backendServer: MockProxy<BackendSrv>;

const mockVariableQueryTestPlansResponse: QueryTestPlansResponse = {
  testPlans: [
    {
      id: '1',
      name: 'testPlan 1',
    },
    {
      id: '2',
      name: 'testPlan 2',
    }
  ],
  totalCount: 2
};

jest.mock('./asset.utils', () => {
  return {
    AssetUtils: jest.fn().mockImplementation(() => ({
      queryAssetsInBatches: jest.fn().mockResolvedValue(
        [
          { id: '1', name: 'Asset 1' },
          { id: '2', name: 'Asset 2' }
        ]
      )
    }))
  };
});

beforeEach(() => {
  [datastore, backendServer] = setupDataSource(TestPlansDataSource);
});

describe('testDatasource', () => {
  test('returns success', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niworkorder/v1/query-testplans', data: { take: 1 } }))
      .mockReturnValue(createFetchResponse(25));

    const result = await datastore.testDatasource();

    expect(result.status).toEqual('success');
  });

  test('bubbles up exception', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niworkorder/v1/query-testplans', data: { take: 1 } }))
      .mockReturnValue(createFetchError(400));

    await expect(datastore.testDatasource())
      .rejects
      .toThrow('Request to url "/niworkorder/v1/query-testplans" failed with status code: 400. Error message: "Error"');
  });

  test('default query output type as properties and default properties', async () => {
    const defaultQuery = datastore.defaultQuery;
    expect(defaultQuery.outputType).toEqual(OutputType.Properties);
    expect(defaultQuery.properties).toEqual([
      Properties.NAME,
      Properties.STATE,
      Properties.ASSIGNED_TO,
      Properties.PRODUCT,
      Properties.DUT_ID,
      Properties.PLANNED_START_DATE_TIME,
      Properties.ESTIMATED_DURATION_IN_SECONDS,
      Properties.SYSTEM_NAME,
      Properties.UPDATED_AT
    ]);
    expect(defaultQuery.orderBy).toEqual(OrderByOptions.UPDATED_AT);
    expect(defaultQuery.descending).toEqual(true);
    expect(defaultQuery.recordCount).toEqual(1000);
  });
});

describe('runQuery', () => {
  const mockOptions: DataQueryRequest = {} as DataQueryRequest;

  test('returns data frame with fields when test plans are available', async () => {
    const query = {
      refId: 'A',
      outputType: OutputType.Properties,
      properties: [Properties.NAME, Properties.STATE],
      orderBy: OrderByOptions.UPDATED_AT,
      recordCount: 10,
      descending: true,
    };

    const testPlansResponse = {
      testPlans: [
        { id: '1', name: 'Test Plan 1', state: 'Active' },
        { id: '2', name: 'Test Plan 2', state: 'Completed' },
      ],
    };

    jest.spyOn(datastore, 'queryTestPlansInBatches').mockResolvedValue(testPlansResponse);

    const result = await datastore.runQuery(query, mockOptions);

    expect(result.fields).toHaveLength(2);
    expect(result.fields[0].name).toEqual('Name');
    expect(result.fields[0].values).toEqual(['Test Plan 1', 'Test Plan 2']);
    expect(result.fields[1].name).toEqual('State');
    expect(result.fields[1].values).toEqual(['Active', 'Completed']);
  });

  test('returns empty data frame when no test plans are available', async () => {
    const query = {
      refId: 'A',
      outputType: OutputType.Properties,
      properties: [Properties.NAME, Properties.STATE],
      orderBy: OrderByOptions.UPDATED_AT,
      recordCount: 10,
      descending: true,
    };

    jest.spyOn(datastore, 'queryTestPlansInBatches').mockResolvedValue({ testPlans: [] });

    const result = await datastore.runQuery(query, mockOptions);

    expect(result.fields).toHaveLength(0);
  });

  test('returns total count when output type is TotalCount', async () => {
    const query = {
      refId: 'A',
      outputType: OutputType.TotalCount,
      orderBy: OrderByOptions.UPDATED_AT,
      recordCount: 10,
      descending: true,
    };

    const testPlansResponse = {
      testPlans: [],
      totalCount: 42,
    };

    jest.spyOn(datastore, 'queryTestPlans').mockResolvedValue(testPlansResponse);

    const result = await datastore.runQuery(query, mockOptions);

    expect(result.fields).toHaveLength(1);
    expect(result.fields[0].name).toEqual('Total count');
    expect(result.fields[0].values).toEqual([42]);
  });

  test('returns empty total count when no test plans are available and output type is TotalCount', async () => {
    const query = {
      refId: 'A',
      outputType: OutputType.TotalCount,
      orderBy: OrderByOptions.UPDATED_AT,
      recordCount: 10,
      descending: true,
    };

    const testPlansResponse = {
      testPlans: [],
      totalCount: 0,
    };

    jest.spyOn(datastore, 'queryTestPlans').mockResolvedValue(testPlansResponse);

    const result = await datastore.runQuery(query, mockOptions);

    expect(result.fields).toHaveLength(1);
    expect(result.fields[0].name).toEqual('Total count');
    expect(result.fields[0].values).toEqual([0]);
  });

  it('should convert fixtureIds to fixture names', async () => {
    const query = {
      refId: 'A',
      outputType: OutputType.Properties,
      properties: [Properties.FIXTURE_NAMES],
      orderBy: OrderByOptions.UPDATED_AT,
      recordCount: 10,
      descending: true,
    };

    const testPlansResponse = {
      testPlans: [
        { id: '1', fixtureIds: ['1'] },
        { id: '2', fixtureIds: ['1', '2'] }
      ],
    };

    jest.spyOn(datastore, 'queryTestPlansInBatches').mockResolvedValue(testPlansResponse);

    const result = await datastore.runQuery(query, mockOptions);

    expect(result.fields).toHaveLength(1);
    expect(result.fields[0].name).toEqual('Fixture names');
    expect(result.fields[0].values).toEqual(['Asset 1', 'Asset 1, Asset 2']);
  });

  it('should convert dutIds to dut names', async () => {
    const query = {
      refId: 'A',
      outputType: OutputType.Properties,
      properties: [Properties.DUT_ID],
      orderBy: OrderByOptions.UPDATED_AT,
      recordCount: 10,
      descending: true,
    };

    const testPlansResponse = {
      testPlans: [
        { id: '1', dutId: '1' },
        { id: '2', dutId: '2' }
      ],
    };

    jest.spyOn(datastore, 'queryTestPlansInBatches').mockResolvedValue(testPlansResponse);

    const result = await datastore.runQuery(query, mockOptions);

    expect(result.fields).toHaveLength(1);
    expect(result.fields[0].name).toEqual('DUT');
    expect(result.fields[0].values).toEqual(['Asset 1', 'Asset 2']);
  });
});

describe('queryTestPlansInBatches', () => {
  test('queries test plans in batches and returns aggregated response', async () => {
    const mockQueryResponse = {
      testPlans: [
        { id: '1', name: 'Test Plan 1' },
        { id: '2', name: 'Test Plan 2' }
      ],
      continuationToken: undefined,
      totalCount: 2,
    };

    jest.spyOn(datastore, 'queryTestPlans').mockResolvedValue(mockQueryResponse);

    const result = await datastore.queryTestPlansInBatches(OrderByOptions.UPDATED_AT, [Projections.NAME], 2, true);

    expect(result.testPlans).toEqual(mockQueryResponse.testPlans);
    expect(result.totalCount).toEqual(2);
  });

  test('handles errors during batch querying', async () => {
    jest.spyOn(datastore, 'queryTestPlans').mockRejectedValue(new Error('Query failed'));

    await expect(datastore.queryTestPlansInBatches(OrderByOptions.UPDATED_AT, [Projections.NAME], 2, true))
      .rejects
      .toThrow('Query failed');
  });
});

describe('queryTestPlans', () => {
  test('sends correct request and returns response', async () => {
    const mockResponse = { testPlans: [{ name: 'Test Plan 1' }], continuationToken: null, totalCount: 1 };

    backendServer.fetch
      .calledWith(requestMatching({ url: '/niworkorder/v1/query-testplans', data: { orderBy: OrderByOptions.UPDATED_AT, take: 1 } }))
      .mockReturnValue(createFetchResponse(mockResponse));

    const result = await datastore.queryTestPlans(OrderByOptions.UPDATED_AT, [Projections.NAME], 1, true);

    expect(result).toEqual(mockResponse);
  });

  test('throws error on failed request', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niworkorder/v1/query-testplans', data: { orderBy: OrderByOptions.UPDATED_AT, take: 1 } }))
      .mockReturnValue(createFetchError(500));

    await expect(datastore.queryTestPlans(OrderByOptions.UPDATED_AT, [Projections.NAME], 1, true))
      .rejects
      .toThrow('An error occurred while querying test plans: Error: Request to url "/niworkorder/v1/query-testplans" failed with status code: 500. Error message: "Error"');
  });
});

describe('metricFindQuery', () => {
  let options: LegacyMetricFindQueryOptions;
  beforeEach(() => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niworkorder/v1/query-testplans' }))
      .mockReturnValue(
        createFetchResponse<QueryTestPlansResponse>(mockVariableQueryTestPlansResponse));
    options = {}
  });

  it('should return test plan name with id when queryBy is not provided', async () => {
    const query: TestPlansVariableQuery = {
      refId: '',
    };

    const results = await datastore.metricFindQuery(query, options);

    expect(results).toMatchSnapshot();
    expect(backendServer.fetch).toHaveBeenCalledWith(
      expect.objectContaining({
        data: {
          descending: false,
          projection: ["ID", "NAME"],
          returnCount: false,
          take: 1000
        }
      })
    );
  });

  it('should return test plan name with id when orderBy is provided', async () => {
    const query: TestPlansVariableQuery = {
      refId: '',
      orderBy: OrderByOptions.UPDATED_AT,
      recordCount: 1000,
      descending: false,
    };

    const results = await datastore.metricFindQuery(query, options);

    expect(results).toMatchSnapshot();
    expect(backendServer.fetch).toHaveBeenCalledWith(
      expect.objectContaining({
        data: {
          descending: false,
          orderBy: OrderByOptions.UPDATED_AT,
          projection: ["ID", "NAME"],
          returnCount: false,
          take: 1000
        }
      })
    );
  });
});
