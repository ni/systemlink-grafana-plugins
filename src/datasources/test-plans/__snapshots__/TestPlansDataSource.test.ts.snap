// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metricFindQuery should return test plan name with id when orderBy is provided 1`] = `
[
  {
    "text": "testPlan 1 (1)",
    "value": "1",
  },
  {
    "text": "testPlan 2 (2)",
    "value": "2",
  },
]
`;

exports[`metricFindQuery should return test plan name with id when queryBy is not provided 1`] = `
[
  {
    "text": "testPlan 1 (1)",
    "value": "1",
  },
  {
    "text": "testPlan 2 (2)",
    "value": "2",
  },
]
`;
